### PR TITLE
Remove "bad characters" from our codebase

### PR DESCRIPTION
**Pull request by potiuk**
We had plenty of "bad characters" in our codebase that were not
invited and came here by accident. We want to get rid of those
"bad characters" once and for all.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Comment by potiuk** [comment]
Turned out we had more "Bad characters" :P in our codebase than literally meets the eye :P

**Comment by potiuk** [comment]
This is no go, I do not want to get ruby dependency for pre-commit. Time for 15-liner python script :)

File: .gitmodules [hunk 1]
```diff
@@ -1,12 +1,12 @@
 [submodule ".github/actions/get-workflow-origin"]
-	path = .github/actions/get-workflow-origin
-	url = https://github.com/potiuk/get-workflow-origin
+    path = .github/actions/get-workflow-origin
```
Comment (ashb) [inline .gitmodules]: Tabs are not bad characters -- flake8/black deals with this, so leave tabs alone.

Comment (ashb) [reply .gitmodules]: Or ignore the following files like we did before:

```
        exclude: ^airflow/_vendor/|^clients/gen/go\.sh$|^\.gitmodules$
```

**Comment by ashb** [comment]
What's wrong with using smart/curly quotes in docs?

File: .gitmodules (hunk 1 above)
Comment (potiuk) [reply .gitmodules]: we alredy had "forbid tabs" before - I just missed an exclusion (which I have to add anyway).

Comment (potiuk) [reply .gitmodules]: (this one will be better because it will fix them automatically).

**Comment by potiuk** [comment]
> What's wrong with using smart/curly quotes in docs?

For the context - this is the result of discussion in #24797  - where *nbsps* were shown in some IDES (only some IDES show it , some others will render it as space.

The main reason is consistency.

If we are using them inconsistently, then it's wrong. You either should use always curly quotes in your docs or always straight ones otherwise our docs will look strange if you look closely.  Another problem it introduces it's "search & replace" functionality. Whenever you are using global search and replace and want to search for "quotes" you will often miss the curly quote cases - especially when it is "random" and not used consistently.

Most of the "curly ones" comes usually (similarly as *nbsp*)  from copy-pasting from word-like sources, but some people will use the "straight" ones, some people will use the "curly" ones and it's difficult to keep consistency.  And the bad thing with it is also that sometimes those curly quotes will sneak-in the code examples and snippets which are part of our documentation and we will not notice. This happened often to me where I wanted to copy some piece of code from the browser, It did not parse/compile or produce strange results. 

Of course there are plenty exceptions where you realy can't use curly ones (code snippets etc.) and it's next to impossible to figure out automatically when yoiu `can` use them.. It's far easier to replace all curly ones  with `straight` ones. The consistency of look&feel is kept and you avoid any of the problems above. And look & feel is marginally impacted IMHO.. And if we REALLY want to keep the curly quotes in some specific files, or folders we can always exclude them by regexp in pre-commit.

**Comment by potiuk** [comment]
Updated much improved version:

1) brought back the same exclusions that "forbid tabs" used to have
2) No ruby/pre-commit dependency - all is pure python + rich (so there is really one virtualenv used for this and many other pre-commits - pre-commit nicely recognizes multiple pre-commits and will reuse the virtualenv that has the same set of requirements.
3) Nicer, colorful  output informing what has been removed (I added "replacements" after the number comparing to the screenshot). 

image

File: scripts/ci/pre_commit/pre_commit_replace_bad_characters.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+import sys
+from pathlib import Path
+from typing import List, NamedTuple
+
+from rich.console import Console
+
+if __name__ != "__main__":
+    raise Exception(
+        "This file is intended to be executed as an executable program. You cannot use it as a module."
+        "To run this script, run the ./build_docs.py command"
```
Comment (ashb) [inline scripts/ci/pre_commit/pre_commit_replace_bad_characters.py]: nit: copy pasta error here.
```suggestion
        f"To run this script, run the {__file__} command"
```

**Review by ashb — APPROVED** [review]

**Pull request merged**